rcl/src/rcl/arguments.c

```diff
@@ -1597,17 +1662,72 @@ rcl_ret_t
 _rcl_parse_log_level(
   const char * arg,
   rcl_allocator_t allocator,
-  int * log_level)
+  rcl_log_levels_t * log_level)
 {
   RCL_CHECK_ARGUMENT_FOR_NULL(arg, RCL_RET_INVALID_ARGUMENT);
   RCL_CHECK_ARGUMENT_FOR_NULL(log_level, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(log_level->logger_settings, RCL_RET_INVALID_ARGUMENT);
+
+  size_t separator_len = strlen(RCL_LOG_LEVEL_FLAG_SEPARATOR);
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+  char * name = NULL;
+  char * level = NULL;
+  int logger_level = 0;
+  char * separator = strstr(arg, RCL_LOG_LEVEL_FLAG_SEPARATOR);
+
+  if (separator) {
+    struct rcl_logger_setting_t * logger_log_level =
+      &log_level->logger_settings[log_level->num_logger_settings];
+    name = rcutils_strndup(arg, separator - arg, allocator);
+    if (name == NULL) {
+      RCL_SET_ERROR_MSG("failed to allocate memory for logger name");
+      return RCL_RET_BAD_ALLOC;
+    }
+    if (strlen(name) == 0) {
+      RCL_SET_ERROR_MSG("Argument has an invalid logger item that name is empty");
+      allocator.deallocate(name, allocator.state);
+      return RCL_RET_ERROR;
+    }
 
-  rcutils_ret_t ret = rcutils_logging_severity_level_from_string(arg, allocator, log_level);
-  if (RCUTILS_RET_OK == ret) {
-    return RCL_RET_OK;
+    level = rcutils_strdup(separator + separator_len, allocator);
+    if (level == NULL) {
+      RCL_SET_ERROR_MSG("failed to allocate memory for logger level");
+      allocator.deallocate(name, allocator.state);
+      return RCL_RET_BAD_ALLOC;
+    }
+    if (strlen(level) == 0) {
+      RCL_SET_ERROR_MSG("Argument has an invalid logger item that level is empty");
+      allocator.deallocate(name, allocator.state);
+      allocator.deallocate(level, allocator.state);
+      return RCL_RET_ERROR;
+    }
+
+    ret = rcutils_logging_severity_level_from_string(
+      level,
+      allocator,
+      &logger_level);
+    allocator.deallocate(level, allocator.state);
+    if (RCUTILS_RET_OK == ret) {
+      logger_log_level->name = name;
+      logger_log_level->level = (rcl_log_severity_t)logger_level;
+      log_level->num_logger_settings += 1;
```

Comment:
consider extending `rcl_log_levels_t` API to handle this:

```c
rcl_ret_t rcl_log_levels_add_logger_setting(const char * logger_name, rcl_log_severity_t log_level);
```

this will make things much easier to unit test.

Reply:
consider adding a `capacity` member to the struct too

Reply:
Updated